src/Microsoft.OData.Core/Metadata/EdmLibraryExtensions.cs

```diff
@@ -252,19 +254,135 @@ internal static IEnumerable<IEdmOperation> FilterBoundOperationsWithSameTypeHier
+        /// <param name="bindingType">Binding type for the operation.</param>
+        /// <param name="parameterNameList">The parameters.</param>
+        /// <param name="caseInsensitive">Whether case insensitive.</param>
+        /// <param name="actionItems">outputs the actions found while matching functions</param>
+        /// <returns>Operations filtered by parameter.</returns>
+        internal static IList<IEdmOperation> FilterOperationCandidatesBasedOnParameterList(this IEnumerable<IEdmOperation> operations, IEdmType bindingType, IList<string> parameterNameList, bool caseInsensitive, out IList<IEdmOperation> actionItems)
+        {
+            Debug.Assert(operations != null, "operations");
+            Debug.Assert(parameterNameList != null, "parameters");
+
+            bool hasParameters = parameterNameList.Count > 0;
+            actionItems = null;
+            IList<IEdmOperation> operationCandidates = new List<IEdmOperation>();
+
+            foreach (var operation in operations)
+            {
+                if (bindingType != null)
+                {
+                    operation.EnsureOperationBoundWithBindingParameter();
+                }
+
+                bool suitableCandidate = true;
+
+                // If the number of parameters > 0 then this has to be a function as actions can't have parameters on the uri only in the payload. Filter further by parameters in this case, otherwise don't.
+                if (hasParameters)
+                {
+                    if (operation.IsAction())
+                    {
+                        if (actionItems == null)
+                        {
+                            actionItems = new List<IEdmOperation>();
+                        }
+
+                        actionItems.Add(operation);
+                        suitableCandidate = false;
+                    }
+
+                    if (!ParametersSatisfyFunction(operation, parameterNameList, caseInsensitive))
+                    {
+                        suitableCandidate = false;
+                    }
+                }
+                else if (bindingType != null)
+                {
+                    // If it is a bound function without required parameters, then the parameter list should only have one parameter and any others as optional.
+                    if (operation.IsFunction())
+                    {
+                        int count = 0;
+                        foreach (var param in operation.Parameters)
+                        {
+                            if (count > 0 && !(param is IEdmOptionalParameter))
+                            {
+                                suitableCandidate = false;
+                            }
+
+                            count++;
+                        }
+
+                        if (count == 0)
+                        {
+                            suitableCandidate = false;
+                        }

```

Comment:
You should be able to break out of the foreach as soon as you find a non-optional parameter.  Can you simplify to something like:

                        foreach (var param in operation.Parameters)
                        {
                            if (!(param is IEdmOptionalParameter))
                            {
                                suitableCandidate = false;
                               break;
                            }
                        }